p2p/proxy.go

```diff
@@ -0,0 +1,92 @@
+package p2p
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+
+	core "github.com/ipfs/go-ipfs/core"
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+	peer "gx/ipfs/QmbNepETomvmXfz1X5pHNFD2QuPqnqi47dTd94QJWSorQ3/go-libp2p-peer"
+)
+
+func ProxyOption() ServeOption {
+	return func(ipfsNode *core.IpfsNode, _ net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
+		mux.HandleFunc("/proxy/", func(w http.ResponseWriter, request *http.Request) {
+			// parse request
+			parsedRequest, err := parseRequest(request)
+			if err != nil {
+				handleError(w, "Failed to parse request", err, 400)
+				return
+			}
+
+			// open connect to peer
+			stream, err := ipfsNode.P2P.PeerHost.NewStream(ipfsNode.Context(), parsedRequest.target, protocol.ID("/x/"+parsedRequest.name))
+			if err != nil {
+				msg := fmt.Sprintf("Failed to open stream '%v' to target peer '%v'", parsedRequest.name, parsedRequest.target)
+				handleError(w, msg, err, 500)
+				return
+			}
+
+			// send request to peer
+			proxyReq, err := http.NewRequest(request.Method, parsedRequest.httpPath, request.Body)
```

Comment:
Does that also work for websockets / multipart requests out of the box?

Reply:
good question, will add a sharness test